source/extensions/filters/udp/dns_filter/dns_parser.cc

```diff
@@ -0,0 +1,305 @@
+                buffer->length());
+      return false;
+    }
+
+    switch (state_) {
+    case DnsQueryParseState::Init:
+      incoming_.id = data;
+      state_ = DnsQueryParseState::Flags;
+      break;
+
+    case DnsQueryParseState::Flags:
+      ::memcpy(static_cast<void*>(&incoming_.flags), &data, sizeof(uint16_t));
+      state_ = DnsQueryParseState::Questions;
+      break;
+
+    case DnsQueryParseState::Questions:
+      incoming_.questions = data;
+      state_ = DnsQueryParseState::Answers;
+      break;
+
+    case DnsQueryParseState::Answers:
+      incoming_.answers = data;
+      state_ = DnsQueryParseState::Authority;
+      break;
+
+    case DnsQueryParseState::Authority:
+      incoming_.authority_rrs = data;
+      state_ = DnsQueryParseState::Authority2;
+      break;
+
+    case DnsQueryParseState::Authority2:
+      incoming_.additional_rrs = data;
+      state_ = DnsQueryParseState::Finish;
+      break;
+
+    case DnsQueryParseState::Finish:
+      break;
+
+    default:
+      NOT_REACHED_GCOVR_EXCL_LINE;
+    }
+  }
+
+  // Verify that we still have available data in the buffer to read answer and query records
+  if (offset > buffer->length()) {
+    ENVOY_LOG(error, "Buffer read offset[{}] is larget than buffer length [{}].", offset,
+              buffer->length());
+    return false;
+  }
+
+  // Each dns request has a Identification ID. This is used to match the request and replies.
+  // We should not see a duplicate ID when handling DNS requests. The ID is removed from the
+  // active transactions queue when we build a response for the identified query
+  const uint16_t id = static_cast<uint16_t>(incoming_.id);
+  if (std::find(active_transactions_.begin(), active_transactions_.end(), id) !=
+      active_transactions_.end()) {
+    ENVOY_LOG(error, "The filter has already encountered ID {} in a previous request", id);
+    return false;
+  }
+
+  // Double check that this ID is not already being handled.
+  if (queries_.find(id) != queries_.end()) {
+    ENVOY_LOG(
+        error,
+        "There are queries matching ID {} from a previous request for which we have not responded",
+        id);
+    return false;
+  }
+
+  active_transactions_.push_back(id);
+
+  // Almost always, we will have only one query here
+  for (auto index = 0; index < incoming_.questions; index++) {
+    ENVOY_LOG(trace, "Parsing [{}/{}] questions", index, incoming_.questions);
+    auto rec = parseDnsQueryRecord(buffer, &offset);
+    if (rec == nullptr) {
+      ENVOY_LOG(error, "Couldn't parse query record from buffer");
+      return false;
+    }
+    storeQueryRecord(std::move(rec));
+  }
+
+  return true;
+}
+
+const std::string DnsMessageParser::parseDnsNameRecord(const Buffer::InstancePtr& buffer,
+                                                       uint64_t* available_bytes,
+                                                       uint64_t* name_offset) {
+  std::stringstream name_ss{};
+  unsigned char c;
+
+  do {
```

Comment:
Have you considered using ares library here that is already linked by envoy?
There are methods to do what you are trying to achieve here - For example "ares_expand_name" does this job for you

Reply:
An example usage is here - https://github.com/sumukhs/envoy-dnsfilter/blob/master/src/dns_codec_impl.cc#L144

Reply:
I did consider it, but didn't see this usage.   Implementing the parsing logic was rather straightforward, IMO.

Reply:
Please do not reimplement the parser if there is a library that can do this. If you do implement the parser it will need a fuzzer.